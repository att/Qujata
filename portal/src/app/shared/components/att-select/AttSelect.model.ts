// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface AttSelectOption<T = any> {
    value: string;
    label: string;
    metadata?: T;
    isDisabled?: boolean,
}

export enum AttSelectTheme {
    REGULAR = 'regular',
    PRIMARY = 'primary',
    WRAPPER = 'wrapper',
}
