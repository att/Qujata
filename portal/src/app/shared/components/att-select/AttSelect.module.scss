@import 'src/styles/variables-keys';
@import './AttSelect.reference';

$attSelectSpinnerSizeValue: var($selectSpinnerSize, 18px);
$attSelectSpinnerRightMarginValue: var($selectSpinnerRightMargin, 10px);
$attSelectSpinnerLeftMarginValue: var($selectSpinnerLeftMargin, 2px);

.att_select_primary_theme {
  #{$selectBorderColor}: var($attCobalt);
  #{$selectControlBackgroundColor}: transparent;
  #{$selectPlaceholderColor}: var($attCobalt);
  #{$selectIndicatorsColor}: var($attCobalt);
  &:hover {
    #{$selectControlBackgroundColor}: var($attCobalt);
    #{$selectPlaceholderColor}: var($primaryWhite);
    #{$selectIndicatorsColor}: var($primaryWhite);
  }
}

.att_select_wrapper_theme {
  #{$selectControlBackgroundColor}: transparent;
  #{$selectBorderColor}: transparent;
}

.att_select_error {
  #{$selectBorderColor}: var($errorTextColor);
}

.att_select_spinner {
  block-size: $attSelectSpinnerSizeValue;
  inline-size: $attSelectSpinnerSizeValue;
  margin-inline-end: $attSelectSpinnerRightMarginValue;
  margin-inline-start: $attSelectSpinnerLeftMarginValue;
}
