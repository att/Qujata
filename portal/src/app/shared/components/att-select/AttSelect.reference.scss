$selectSpinnerSize: --att-select-spinner-size;
$selectSpinnerRightMargin: --att-select-spinner-right-margin;
$selectSpinnerLeftMargin: --att-select-spinner-left-margin;
