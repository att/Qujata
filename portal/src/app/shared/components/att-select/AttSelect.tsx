/* eslint-disable react/no-unused-prop-types */
/* eslint-disable @typescript-eslint/ban-types */
import cn from 'classnames';
import React, { FocusEventHandler, ForwardedRef, forwardRef, ForwardRefExoticComponent, PropsWithChildren, PropsWithoutRef, RefAttributes } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import Select, { ActionMeta, Options, SelectInstance } from 'react-select';
import styles from './AttSelect.module.scss';
import { AttSelectOption, AttSelectTheme } from './AttSelect.model';
import { Spinner, SpinnerSize } from '../att-spinner';
import { AttSelectCssClassByTheme } from './AttSelect.const';

export type OnSelectChangedCallbackType = (newValue: unknown, actionMeta: ActionMeta<unknown>) => void;
export type OnSelectChangedType = AttSelectOption | Options<AttSelectOption> | null;
export type OnSelectChanged = (event: OnSelectChangedType) => void;

export interface AttSelectProps {
  options: AttSelectOption[];
  value: AttSelectOption | AttSelectOption[] | undefined | null;
  required?: boolean;
  defaultValue?: AttSelectOption | AttSelectOption[];
  onChange: OnSelectChanged;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  isProcessing?: boolean;
  disabled?: boolean;
  readonly?: boolean;
  isClearable?: boolean;
  hasError?: boolean;
  className?: string;
  placeholder?: string;
  customComponent?: {
    Option?: React.FC;
    ValueContainer?: React.FC;
    DropdownIndicator?: React.FC;
    MenuList?: React.FC;
  };
  theme?: AttSelectTheme;
  isMulti?: boolean;
  closeMenuOnSelect?: boolean;
  hideSelectedOptions?: boolean;
  e2eId?: string;
  id?: string;
}

export type AttSelectExposeRef = SelectInstance;
type AttSelectPrivateType = React.ForwardRefRenderFunction<AttSelectExposeRef, AttSelectProps>;
const AttSelectPrivate: AttSelectPrivateType = (props: AttSelectProps, ref: ForwardedRef<AttSelectExposeRef>) => (
  <Select
    ref={ref}
    className={getAttSelectClassName(props)}
    classNamePrefix='att_select'
    options={props.options}
    onChange={props.onChange as OnSelectChangedCallbackType}
    onBlur={props.onBlur}
    isDisabled={props.disabled || props.readonly}
    isLoading={props.isProcessing}
    defaultValue={props.defaultValue}
    value={props.value}
    placeholder={props.placeholder}
    isMulti={props.isMulti}
    isClearable={props.isClearable}
    // menuIsOpen={true}
    closeMenuOnSelect={props.closeMenuOnSelect}
    hideSelectedOptions={props.hideSelectedOptions}
    required={props.required}
    components={{
      IndicatorSeparator,
      LoadingIndicator,
      ...(props.customComponent || {}),
    }}
    id={props.e2eId}
    inputId={props.id}
  />
);

function getAttSelectClassName(props: AttSelectProps): string {
  const { className, readonly, isProcessing, hasError, isMulti } = props;
  const theme: AttSelectTheme = props.theme || AttSelectTheme.REGULAR;
  return cn(
    'att_select_container',
    className,
    AttSelectCssClassByTheme.get(theme),
    {
      'att_select_container--is-readonly': readonly,
      'att_select_container--is-loading': isProcessing,
      att_select_multi_container: isMulti,
      [styles.att_select_error]: hasError,
    },
  );
}
function IndicatorSeparator() {
  // eslint-disable-next-line no-null/no-null
  return null;
}

function LoadingIndicator() {
  return <Spinner size={SpinnerSize.NONE} className={styles.att_select_spinner} />;
}

type AttSelectRef = ForwardRefExoticComponent<PropsWithoutRef<PropsWithChildren<AttSelectProps>> & RefAttributes<AttSelectExposeRef>>
export const AttSelect: AttSelectRef = forwardRef<AttSelectExposeRef, PropsWithChildren<AttSelectProps>>(AttSelectPrivate);
