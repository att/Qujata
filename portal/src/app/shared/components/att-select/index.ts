export * from './AttSelect';
export * from './AttSelect.model';
