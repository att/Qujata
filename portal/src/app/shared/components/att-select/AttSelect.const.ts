import styles from './AttSelect.module.scss';
import { AttSelectTheme } from './AttSelect.model';

export const AttSelectCssClassByTheme: Map<AttSelectTheme, string> = new Map<AttSelectTheme, string>([
  [AttSelectTheme.PRIMARY, styles.att_select_primary_theme],
  [AttSelectTheme.WRAPPER, styles.att_select_wrapper_theme],
  [AttSelectTheme.REGULAR, ''],
]);
