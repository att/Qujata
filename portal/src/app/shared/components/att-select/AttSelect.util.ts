import { OptionProps } from 'react-select';
import { AttSelectOption } from './AttSelect.model';

export function getSelectOptionClassName({ cx, isDisabled, isFocused, isSelected, data }: OptionProps<AttSelectOption, boolean>): string {
  return cx(
    {
      option: true,
      // from some reason in production build isDisabled is coming with wrong value, to insure it we will check also isDisabled value in data object
      // this is the same place react-select is taking isDisabled value (see: builtins.ts file -> isOptionDisabled function)
      // react-select expose isOptionDisabled function which return true/false if the option should be disabled, it's getting also option object (our data object)
      'option--is-disabled': isDisabled || !!data.isDisabled,
      'option--is-focused': isFocused,
      'option--is-selected': isSelected,
    },
  );
}
