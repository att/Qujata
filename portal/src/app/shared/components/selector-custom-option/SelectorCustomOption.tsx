import { OptionProps, components } from 'react-select';
import styles from './SelectorCustomOption.module.scss';
import { AttSelectOption } from '../att-select';
import { algorithmSections } from '../../../components/protocol-query/constants';
import cn from 'classnames';
import CheckedSvg from '../../../../assets/images/checked.svg';
import UnCheckedSvg from '../../../../assets/images/unchecked.svg';

export type SelectorCustomOptionProps = OptionProps<AttSelectOption, true> & {
  onOptionChanged: (option: AttSelectOption) => void;
};

export const AlgorithmsSelectorCustomOption: React.FC<SelectorCustomOptionProps> = (props: SelectorCustomOptionProps) => {
  const isSectionTitle = algorithmSections.includes((props.data as AttSelectOption).value);
  const optionStyle = isSectionTitle ? styles.algorithms_input_option_title : styles.algorithms_input_option;

  return (
    <components.Option {...props}>
      <div className={styles.option_wrapper}>
        <input
          type="checkbox"
          className={styles.input_option}
          checked={props.isSelected}
          onChange={() => props.onOptionChanged} />
          <img className={optionStyle} src={props.isSelected ? CheckedSvg : UnCheckedSvg} alt="checked" />
      </div>
      <span className={optionStyle}>{props.label}</span>
    </components.Option>
  );
};

export const IterationsSelectorCustomOption: React.FC<SelectorCustomOptionProps> = (props: SelectorCustomOptionProps) => {
  return (
    <components.Option {...props}>
      <div className={styles.option_wrapper}>
        <input
          type="checkbox"
          className={cn(styles.iterations_input_option, styles.input_option)}
          checked={props.isSelected}
          onChange={() => props.onOptionChanged} />
          <img src={props.isSelected ? CheckedSvg : UnCheckedSvg} alt="checked" />
      </div>
      <span>{props.label}</span>
    </components.Option>
  );
};