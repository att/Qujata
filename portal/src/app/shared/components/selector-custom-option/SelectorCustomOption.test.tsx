import { fireEvent, render, RenderResult, waitFor } from '@testing-library/react';
import { AlgorithmsSelectorCustomOption, SelectorCustomOption, SelectorCustomOptionProps } from './SelectorCustomOption';
import { AttSelectOption } from '../att-select';

jest.mock('../../../components/protocol-query/constants', () => ({
  algorithmSections: ['All', 'Classic', 'Hybrid', 'PQ']
}));

describe('SelectorCustomOption', () => {
  let mockOption: AttSelectOption;
  let mockProps: SelectorCustomOptionProps;

  beforeEach(() => {
    mockOption = { value: 'Classic', label: 'Classic' };

    mockProps = {
      data: mockOption,
      isSelected: false,
      selectOption: jest.fn(),
      label: 'Option 1',
      innerProps: {},
      innerRef: jest.fn(),
      children: null,
      type: 'option',
      isDisabled: false,
      isFocused: false,
      clearValue: jest.fn(),
      cx: jest.fn(),
      getStyles: jest.fn(),
      getClassNames: jest.fn(),
      getValue: jest.fn().mockReturnValue([{ label: 'Option 1', value: 'option1' }]),
      hasValue: true,
      isMulti: true,
      isRtl: false,
      options: [],
      selectProps: expect.any(Object),
      setValue: jest.fn(),
      theme: expect.any(Object),
      onOptionChanged: jest.fn(),
      showInputOption: false,
      setShowInputOption: jest.fn(),
      inputValue: '1111',
      setInputValue: jest.fn(),
      setMenuIsOpen: jest.fn(),
    };
  });

  test('should render AlgorithmsSelectorCustomOption correctly', async () => {
    const { container, getByTestId }: RenderResult = render(<AlgorithmsSelectorCustomOption {...mockProps} />);
    const algorithmCheckboxElement: HTMLElement = getByTestId('algorithm_checkbox');

    fireEvent.click(algorithmCheckboxElement);

    await waitFor(() => {
      expect(container).toBeTruthy();
    });
  });

  test('should render AlgorithmsSelectorCustomOption correctly when the prop isSelected', async () => {
    mockProps.isSelected = true;
    mockProps.data = { value: 'non-title', label: 'non-title' };
    const { container, getByTestId }: RenderResult = render(<AlgorithmsSelectorCustomOption {...mockProps} />);
    const algorithmCheckboxElement: HTMLElement = getByTestId('algorithm_checkbox');

    fireEvent.click(algorithmCheckboxElement);

    await waitFor(() => {
      expect(container).toBeTruthy();
    });
  });

  test('should render SelectorCustomOption correctly', async () => {
    const { container, getByTestId }: RenderResult = render(<SelectorCustomOption {...mockProps} />);
    const algorithmCheckboxElement: HTMLElement = getByTestId('selector_custom_option_checkbox');

    fireEvent.click(algorithmCheckboxElement);

    await waitFor(() => {
      expect(container).toBeTruthy();
    });
  });

  test('should render SelectorCustomOption correctly when the prop isSelected', async () => {
    mockProps.isSelected = true;
    mockProps.data = { value: 'non-title', label: 'non-title' };
    const { container, getByTestId }: RenderResult = render(<SelectorCustomOption {...mockProps} />);
    const algorithmCheckboxElement: HTMLElement = getByTestId('selector_custom_option_checkbox');

    fireEvent.click(algorithmCheckboxElement);

    await waitFor(() => {
      expect(container).toBeTruthy();
    });
  });
});