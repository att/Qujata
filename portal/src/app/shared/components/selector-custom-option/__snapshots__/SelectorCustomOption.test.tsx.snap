// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectorCustomOption should render AlgorithmsSelectorCustomOption correctly 1`] = `
<div
  aria-disabled="false"
  class="css-0"
>
  <div
    class="option_wrapper"
  >
    <input
      class="input_option"
      id="Option 1"
      type="checkbox"
    />
    <img
      alt="checked"
      class="algorithms_input_option"
      src="unchecked.svg"
    />
  </div>
  <span
    class="algorithms_input_option"
  >
    Option 1
  </span>
</div>
`;

exports[`SelectorCustomOption should render SelectorCustomOption correctly 1`] = `
<div
  aria-disabled="false"
  class="css-0"
>
  <div
    class="option_wrapper"
  >
    <input
      class="custom_input_option input_option"
      id="Option 1"
      type="checkbox"
    />
    <img
      alt="checked"
      src="unchecked.svg"
    />
  </div>
  <span>
    Option 1
  </span>
</div>
`;
